tests/isa-test.cc

```diff
@@ -0,0 +1,774 @@
+			                     Permission::Load,
+			                     Permission::Store,
+			                     Permission::LoadStoreCapability,
+			                     Permission::LoadGlobal,
+			                     Permission::LoadMutable,
+			                     Permission::StoreLocal} &
+			       ps;
+		}
+		if (ps.contains(Permission::Load, Permission::LoadStoreCapability))
+		{
+			// cap-ro format
+			return PermissionSet{Permission::Global,
+			                     Permission::Load,
+			                     Permission::LoadStoreCapability,
+			                     Permission::LoadGlobal,
+			                     Permission::LoadMutable} &
+			       ps;
+		}
+		if (ps.contains(Permission::Store, Permission::LoadStoreCapability))
+		{
+			// cap-wo format
+			return PermissionSet{Permission::Global,
+			                     Permission::Store,
+			                     Permission::LoadStoreCapability} &
+			       ps;
+		}
+		if (ps.contains(Permission::Store) || ps.contains(Permission::Load))
+		{
+			// data-rw format
+			return PermissionSet{
+			         Permission::Global, Permission::Load, Permission::Store} &
+			       ps;
+		}
+		// sealing format
+		return PermissionSet{Permission::Global,
+		                     Permission::Seal,
+		                     Permission::Unseal,
+		                     Permission::User0} &
+		       ps;
+	}
+
+	template<typename T>
+	void test_restrict_capability(Capability<T> capability, PermissionSet mask)
+	{
+		PermissionSet requestedPermissions(mask);
+		requestedPermissions &= capability.permissions();
+		PermissionSet expectedPermissions =
+		  permissions_to_representable(requestedPermissions);
+		capability.permissions() &= mask;
+		TEST(capability.permissions() == expectedPermissions,
+		     "permissions {} did not match expected {}",
+		     capability.permissions(),
+		     expectedPermissions);
+	}
+
+	void test_and_perms()
+	{
+		int              myLocal = 0;
+		const Capability CapabilityToLocal{&myLocal};
+		const Capability CapabilityToGlobal{myGlobalIntPointer};
+		const Capability CapabilityToFunction{
+		  reinterpret_cast<const void *>(&test_and_perms)};
+		debug_log("Checking and perms results...");
+		// TODO we could do with a sealing capability too, but then this would
+		// need to become a privileged compartment.
+		// Test all possible masks on our 'root' capabilities. For now we only
+		// test permissions on data / executable capabilities. Conveniently
+		// these occupy the lower 9 bits of the permissions field.
+		for (uint32_t p = 0; p <= 0x1ff; p++)
+		{
+			auto permissionMask = PermissionSet::from_raw(p);
+			test_restrict_capability(CapabilityToLocal, permissionMask);
+			test_restrict_capability(CapabilityToGlobal, permissionMask);
+			test_restrict_capability(CapabilityToFunction, permissionMask);
+		}
+	}
+
+	[[cheri::interrupt_state(disabled)]] void *get_interrupts_disabled_sentry()
+	{
+		void *returnCap;
+		asm volatile("cjal %0, 4" : "=C"(returnCap));
```

Comment:
This is beautifully horrifying, can you add a comment explaining it?

Reply:
It is just a jump and link to the next instruction, putting the link cap in `returnCap`. I guess might be better written with local label rather than hardcoded offset?